AWSClientRuntime/Sources/Middlewares/Sha256TreeHashMiddleware.swift

```diff
@@ -0,0 +1,107 @@
+    where H: Handler,
+          Self.MInput == H.Input,
+          Self.MOutput == H.Output,
+          Self.Context == H.Context,
+          Self.MError == H.MiddlewareError {
+              let request = input.build()
+              
+              switch request.body {
+              case .data(let data):
+                  guard let data = data else {
+                      return next.handle(context: context, input: input)
+                  }
+                  if !request.headers.exists(name: X_AMZ_CONTENT_SHA256_HEADER_NAME) {
+                      let sha256 = ByteBuffer(data: data).sha256().encodeToHexString()
+                      input.withHeader(name: X_AMZ_CONTENT_SHA256_HEADER_NAME, value: sha256)
+                  }
+              case .stream(let stream):
+                  let streamBytes = stream.toBytes()
+                  guard streamBytes.length > 0 else {
+                      return next.handle(context: context, input: input)
+                  }
+                  let (linearHash, treeHash) = computeHashes(bytes: streamBytes)
+                  if let treeHash = treeHash, let linearHash = linearHash {
+                      input.withHeader(name: X_AMZ_SHA256_TREE_HASH_HEADER_NAME, value: treeHash)
+                      input.withHeader(name: X_AMZ_CONTENT_SHA256_HEADER_NAME, value: linearHash)
+                  }
+              case .empty, .none:
+                  break
+              }
+              
+              return next.handle(context: context, input: input)
+          }
+    
+    /// Computes the tree-hash and linear hash of a `ByteBuffer`.
+    /// See http://docs.aws.amazon.com/amazonglacier/latest/dev/checksum-calculations.html for more information.
+    private func computeHashes(bytes: ByteBuffer) -> (String?, String?) {
+        let bufferSize = 1024 * 1024
+        var hashes = [[UInt8]]()
+        
+        while true {
+            var oneMbTempBuffer = ByteBuffer(size: bufferSize)
+            let bytesRead = bytes.readIntoBuffer(buffer: &oneMbTempBuffer)
+            if bytesRead == 0 {
+                break
+            }
+            let hash = oneMbTempBuffer.sha256()
+            hashes.append(hash.toByteArray())
+        }
+        
+        return (bytes.sha256().encodeToHexString(), computeTreeHash(hashes: &hashes))
+    }
+    
+    /// Builds a tree hash root node given a slice of hashes. Glacier tree hash to be derived from SHA256 hashes of 1MB chunks of the data.
+    /// See http://docs.aws.amazon.com/amazonglacier/latest/dev/checksum-calculations.html for more information.
+    private func computeTreeHash(hashes: inout [[UInt8]]) -> String? {
+        guard !hashes.isEmpty else {
+            return nil
+        }
+        
+        while hashes.count > 1 {
+            var tempHashes = [[UInt8]]()
+            for index in stride(from: 0, to: hashes.count, by: 2) {
+                if index + 1 <= hashes.count - 1 {
```

Comment:
is it possible to do `hashes.count - index > 1` so that we mirror the example?

Reply:
updated